### DISCUSSION TITLE
3.1.0 Participant user data/get_matched_publication_data example

### DISCUSSION THREAD

**Discussion by CaseySanchez** [eProsima/Fast-DDS#5409] [Q&A] [answered]
Hello eProsima and the Fast-DDS community,

Are there any examples that demonstrate how to use participant user data?

**Description:**
To represent my user data I have created a `TestUserData.idl`, which is then assigned to the `DomainParticipantQos` of the participant publisher in my `TestPublisher` class. I create a participant subscriber to this topic in my `TestSubscriber` class, which in turn calls the `get_matched_publication_data` method in the `on_subscription_matched` callback. This simplified example code can be seen below.

**TestUserData.idl**
```
struct TestUserData
{
    unsigned long source_id;
};
```

**TestPublisher.cpp**
```
TestPublisher::TestPublisher(uint32_t const &source_id)
{
    DomainParticipantQos test_participant_qos;

    test_participant_qos.name("TestParticipant");

    eprosima::fastdds::dds::xtypes::SBoundSeq test_user_data_seq(sizeof(TestUserData));

    TestUserData *test_user_data = reinterpret_cast<TestUserData *>(test_user_data_seq.data());

    test_user_data->source_id() = source_id;

    test_participant_qos.user_data().setValue(test_user_data_seq);

    m_test_participant = DomainParticipantFactory::get_instance()->create_participant(0, test_participant_qos);

    if (m_test_participant == nullptr) {
        throw std::runtime_error("Failed to create domain participant.");
    }

    // Register the Type
    m_test_type.register_type(m_test_participant, "TestMessage");

    // Create the subscriptions Topic
    m_test_topic = m_test_participant->create_topic("TestTopic", "TestMessage", TOPIC_QOS_DEFAULT);

    if (m_test_topic == nullptr) {
        throw std::runtime_error("Failed to create topic.");
    }

    // Create the Publisher
    m_test_publisher = m_test_participant->create_publisher(PUBLISHER_QOS_DEFAULT, nullptr);

    if (m_test_publisher == nullptr) {
        throw std::runtime_error("Failed to create publisher.");
    }

    // Create the DataWriter
    m_test_writer = m_test_publisher->create_datawriter(m_test_topic, DATAWRITER_QOS_DEFAULT, this);

    if (m_test_writer == nullptr) {
        throw std::runtime_error("Failed to create data writer.");
    }

    m_test_thread = std::thread(&TestTask::test_thread_handler, this);
}
```

**TestSubscriber.cpp**
```
void TestSubscriber::on_subscription_matched(DataReader *data_reader, SubscriptionMatchedStatus const &subscription_matched_status)
{
    if (subscription_matched_status.current_count_change == 1) {
        PublicationBuiltinTopicData publication_data;

        if (data_reader->get_matched_publication_data(publication_data, subscription_matched_status.last_publication_handle) != eprosima::fastdds::dds::RETCODE_OK) {
            return;
        }

        if (publication_data.user_data.getValue().size() != sizeof(TestUserData)) {
            return;
        }

        TestUserData const *test_user_data = reinterpret_cast<TestUserData const *>(publication_data.user_data.getValue().data());

        // ...

        std::cout << "Subscriber matched." << std::endl;
    }
    else if (subscription_matched_status.current_count_change == -1) {
        PublicationBuiltinTopicData publication_data;

        if (data_reader->get_matched_publication_data(publication_data, subscription_matched_status.last_publication_handle) != eprosima::fastdds::dds::RETCODE_OK) {
            return;
        }

        if (publication_data.user_data.size() != sizeof(TestUserData)) {
            return;
        }

        TestUserData const *test_user_data = reinterpret_cast<TestUserData const *>(publication_data.user_data.getValue().data());

        // ...

        std::cout << "Subscriber unmatched." << std::endl;
    }
    else {
        std::cout << subscription_matched_status.current_count_change << " is not a valid value for SubscriptionMatchedStatus current count change" << std::endl;
    }
}
```

**Problem:**
The call to `get_matched_publication_data` in turns calls `matched_writer_is_matched`, which attempts to acquire a lock on a mutex as follows: `std::lock_guard<RecursiveTimedMutex> guard(mp_mutex);`. It is unable to do so and hangs indefinitely.

Are there any suggestions to using this new feature?

Regards,
Casey

**Comment by Mario-DL** [comment, marked as answer]
Hi @CaseySanchez,

Thanks for using `Fast DDS`. I would clarify some ideas here:
* The [UserData](https://fast-dds.docs.eprosima.com/en/latest/fastdds/dds_layer/core/policy/standardQosPolicies.html#userdataqospolicy) is meant for the user to add extra information of the entity (vector of bytes), but you usually use a standard topic for transmitting a data message starting from an `idl`.  
* A usual way of retrieving the `UserDataQoSPolicy` in the receiving side would be in the `on_participant_discovery()` within `ParticipantDiscoveryInfo`.
* Making operations on discovery callbacks is usually discouraged as it can potentially cause deadlocks with the application.

**Reply by CaseySanchez** [reply]
Thank you kindly @Mario-DL, assigning to the user data in the `DataWriterQos` struct on the publisher side and moving the logic into `on_data_writer_discovery()` on the subscriber side worked for my use-case.

Regarding list item 1, the `TestUserData.idl` was built with the `-no-typesupport` of `fastddsgen`. The intent was to provide a structure to be cast to/from rather than manipulating the raw bytes.

Marked as solved.
